### ISSUE TITLE
Zipping multiple streams is not possible to do naively

### ISSUE THREAD

**Issue by bojidar-bg** [obecto/perper#69]
Zipping multiple streams, either with `System.Linq.Async` or manually with `GetAsyncEnumerator` / `MoveNextAsync` doesn't work within another stream due to the ordering of notifications enforced by Fabric.

**Reproduction:**

```csharp
[FunctionName("TestZip")]
public static async IAsyncEnumerable<(object, object)> TestZip(
  [PerperTrigger] (IAsyncEnumerable<object>, IAsyncEnumerable<object>) streams)
{
  var enumerators = (streams.Item1.GetAsyncEnumerator(), streams.Item2.GetAsyncEnumerator());
  while (await enumerators.Item1.MoveNextAsync() && await enumerators.Item2.MoveNextAsync())
  {
    var items = (enumerators.Item1.Current, enumerators.Item2.Current);
    yield return items;
  }
}
```

While this code seems like it should be working (it just iterates two streams together, zipping them), in practice this does not produce any output.

**Workarounds:**

It is possible to workaround the issue by listening the streams in parallel and keeping track of unprocessed items in state. However, the code for this is somewhat lengthy.

<details>

```csharp
[FunctionName("TestZip")]
public static async IAsyncEnumerable<(object, object)> TestZip(
  [PerperTrigger] (IAsyncEnumerable<object>, IAsyncEnumerable<object>) streams,
  IStateEntry<(Queue<object>, Queue<object>)> state)
{
  var channel = Channel.CreateUnbounded<(object, object)>;
  async Task UpdateState()
  {
    while(state.Value.Item1.Count > 1 && state.Value.Item2.Count > 1)
    {
      await channel.Writer.WriteAsync((state.Value.Item1.Dequeue(), state.Value.Item2.Dequeue()));
    }
  }
  using var cancellation = new CancellationTokenSource();
  var enumerationTask = Task.WhenAll(
    Task.Run(async () =>
    {
      await foreach(var item in streams.Item1.WithCancellation(cancellation.Token))
      {
        state.Value.Item1.Enqueue(item);
        await UpdateState();
      }
    }),
    Task.Run(async () =>
    {
      await foreach(var item in streams.Item2.WithCancellation(cancellation.Token))
      {
        state.Value.Item2.Enqueue(item);
        await UpdateState();
      }
    }));
  await foreach (var item in channel.Reader.ReadAllAsync())
  {
    yield return item;
  }
  cancellation.Cancel();
  await enumerationTask;
}
```

</details>

**Comment by bojidar-bg** [comment]
**Debugging:**

Assuming perfect timing between the two parameter streams, the notifications queue for `TestZip` looks something like this after the two streams have sent their notifications:
```
0: StreamItemNotification { stream="TestZip-GUID", parameter=1, .. }
1: StreamItemNotification { stream="TestZip-GUID", parameter=2, .. }
2: StreamItemNotification { stream="TestZip-GUID", parameter=1, .. }
3: StreamItemNotification { stream="TestZip-GUID", parameter=2, .. }
..
```

However, due to the constraints enforced by Fabric, we cannot receive notification `1` before processing notification `0`. Thus, the second `MoveNextAsync` call will hang indefinitelly, as we have not consumed the notification resulting from the first `MoveNextAsync` call.

The only way in which the reproduction code above would work is if we separate the notification queues by parameter index. (Which happens to be similar to what the workaround code does)

**Comment by bojidar-bg** [comment]
Was worked around in 7a24f844439962174dbbab19c68b612fb84eac34 and later fixed in 07e632362cc62f0c2d428b5bf29b8d0768d0b61a.
Due to new listeners mechanism, in 0.8 it should be possible to directly `.Zip` streams and not even worry for issues related to state.
